esmvaltool/preprocessor/_multimodel.py

```diff
@@ -46,9 +199,114 @@ def _slice_cube(cube, min_t, max_t):
     return cube_slice
 
 
+def _slice_cube2(cube, t_1, t_2):
+    """
+    Efficient slicer
+
+    Slice cube on time more memory-efficiently than
+    _slice_cube(); using this function adds virtually
+    no extra memory to the process
+    """
+    time_units = cube.coord('time').units
+    time_pts = [j for j in cube.coord('time').points]
+    if time_units.calendar == '360_day':
+        # apply  calendar unifying prefactor
+        a_p = 365. / 360.
+    else:
+        a_p = 1.0
+    idxs = sorted([time_pts.index(ii)
+                   for ii in time_pts
+                   if t_1 <= ii * a_p <= t_2])
+    cube_t_slice = cube.data[idxs[0]:idxs[-1] + 1]
+    return cube_t_slice
+
+
+def _monthly_t(cubes):
+    """Rearrange time points for monthly data"""
+    # get original cubes tpoints
+    tpts = []
+    # make all 365-day calendards
+    for cube in cubes:
+        time_units = cube.coord('time').units
+        if time_units.calendar is not None:
+            if time_units.calendar == '360_day':
+                a_p = 365. / 360.
+            else:
+                a_p = 1.
+        else:
+            logger.debug("No calendar type: assuming 365-day")
+            a_p = 1.
+        tpts.append([int(c * a_p) for c in cube.coord('time').points])
+
+    # convert to months for MONTHLY data
+    t_x = list(set().union(*tpts))
+    t_x = list(set([int((a / 365.) * 12.) for a in t_x]))
+    t_x.sort()
+    # remake the time axis for roughly the 15th of the month
+    t_x0 = list(set([t * 365. / 12. + 15. for t in t_x]))
+    t_x0.sort()
+
+    return t_x, t_x0
+
+
+def _full_time(cubes):
+    """Construct a contiguous collection over time"""
+    datas = []
+    # get rearranged time points
+    t_x, t_x0 = _monthly_t(cubes)
+    # loop through cubes and apply masks
+    for cube in cubes:
+        # recast time points
```

Comment:
Make a function to convert time coordinates to time points with the correct calendar, you have more or less the same code to do this 4 times in this module.

Reply:
done